### ISSUE TITLE
Transactional flow duplicates inflight messages during rebalancing

### ISSUE THREAD

**Issue by szymonm** [akka/alpakka-kafka#758]
### Versions used 
Alpakka Kafka version: master

### Expected Behavior

Transactional flow should not duplicate messages when partitions are rebalanced by Kafka.

### Actual Behavior

Messages that are in-flight during rebalancing are duplicated.

### Reproducible Test Case

See #757 for a test case.

**Comment by szymonm** [comment]
Probably related to  #756

**Comment by szymonm** [comment]
It seems to me that to fix this issue we should implement a logic similar to what we have in `CommittableSource`. Precisely, when partitions are revoked all consumers should wait for all the in-flight records to be committed before calling to retrieve offsets from the newly assigned partitions.

To achieve this I suggest:
1. Keep state of all committed offsets in `KafkaConsumerActor` to know whether records emitted from the `KafkaConsumerActor` are no longer in the stream.
2. Add waiting for flushing all records in the `onRevoke` `ListenerCallback`.
3. Add a way of signalling that an offset was successfully committed after `    producer.commitTransaction()` in `TransactionalProducerStage` to the `KafkaConsumerActor`. Probably by extending `TransactionalMessage`with some committer equivalent.

What do you think @seglo @ennru @2m ?

**Comment by 2m** [comment]
Yes, this is the same issue as is described in #756 where the rebalance is triggered by restarting one of the consumers from the consumer group.

Action plan sounds good. I wonder if `CommittableSource` could be reused for TransactionalSource but it's commit initiation logic would be replaced with the one coming from `TransactionalProducerStage`, as you suggested.

**Comment by szymonm** [comment]
Is anyone of you working currently on this issue, @2m ?  We need this to be fixed, so I'm ready to work on this for some time, but I don't want to duplicate your work, if you already have some attempts to solve it...

**Comment by 2m** [comment]
I have been working on this mostly to understand what is going on under the hood. Please go ahead and make the required changes. I will make sure to give speedy feedback so we can fix this sooner rather than later.

**Comment by seglo** [comment]
@szymonm Sounds good to me.  I've been thinking about this problem over the weekend and I agree that using the callback listener is probably required.  I'm looking forward to see what you come up with :)

**Comment by ennru** [comment]
I've tried to simplify the actor's rebalance listener in #753 (merged) and #759. The commit tracking for commit refreshing might be useful for this problem here, as well.

I'll look again at how we could hand out a blocking rebalance listener to the library user #539.
